### ISSUE TITLE
Editable mode is gone since `1.1.0a2`

### ISSUE THREAD

**Issue by snejus** [python-poetry/poetry#2725]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.  
  * Though this covers more than one release
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: Ubuntu 18.04.4 LTS x86_64
- **Poetry version**: `1.1.0a2-b2`
- **Link of a [Gist](https://gist.github.com/) with the contents of your pyproject.toml file**: 

## Issue

I've been using `1.1.0a1` version since the beginning and having seen the recent pre-release updates I thought I'd try them out. Unfortunately, it doesn't seem that Poetry has been able to install *the root project* in editable mode since `1.1.0a2` pre-release. I've checked also checked `a3` and `b2` and found the same behaviour.

`pip list` output using `1.1.0a1`, as expected:
```
requests-toolbelt     0.9.1
safety                1.9.0
my-package 4.4.0     /projects/my-package/src
```

and what I saw for the more recent pre-releases mentioned above:
```
requests-toolbelt     0.9.1
safety                1.9.0
my-package 4.4.0     /projects/my-package/.venv/lib/python3.6/site-packages
```

It's telling me that the editable installation is in the `site-packages` - by this logic every each of my dependencies is an editable one too :sweat_smile: 

Would be good to have sorted since it's something that's blocking me from updating. I'd be happy to help with this if you need extra hands.

See below for the `pyproject.toml`. The package metadata is altered a bit
```toml
[tool.poetry]
name = "my-package"
version = "4.4.0"
authors = ["Sarunas Nejus <my@email.com>"]
description = "My cool cool package"

readme = "README.rst"
repository = "https://github.com/my/package"

packages = [{ include = "mypackage", from = "src" }]
include = ["packagedata/**/*"]

[tool.poetry.urls]
Changelog = "https://github.com/my/package/CHANGELOG.md"

[[tool.poetry.source]]
name = "pripy"
url = "https://hidden-pypi.com"

[tool.poetry.dependencies]
python = "~3.6"

pydantic = "^1.5"
dataclasses = "^0.7"
dj-database-url = "^0.4"
psycopg2-binary = "^2.8"
python-dateutil = "*"
typing-extensions = "*"

[tool.poetry.dev-dependencies]
mypy = "^0.782"
flake8 = "^3.8"
safety = "^1.9"
pytest = "^5.4"
pylint = "^2.5"
pytest-cov = "^2.8"
pytest-randomly = "^3"


[tool.coverage.run]
branch = true
dynamic_context = "test_function"

[tool.coverage.report]
precision = 2
show_missing = true
exclude_lines = [
    "pragma: no cover",
    "if TYPE_CHECKING",
    "raise AssertionError",
    "raise NotImplementedError"
]

[tool.coverage.html]
show_contexts = true

[tool.pylint.MASTER]
class-rgx = "[A-Z_]+[a-zA-Z0-9]+$"
deprecated-modules = ["tkinter.tix", "formatter", "pycrypto", "optparse", "httpoxy", "xmlrpc",
                      "parser", "email", "pipes", "crypt", "spwd", "imp", "xml", "lxml"]
disable = [
    "no-member",  # buggy
    "bad-continuation",  # buggy
    "ungrouped-imports",  # incompatible with mypy
    "no-name-in-module",  # buggy
    "undefined-variable",  # covered by mypy
    "not-callable",  # covered by mypy
    "unused-import", # covered by flake8 which does it correctly
]
docstring-min-length = 10
good-names = "_,pk,pytest_plugins"
ignore-long-lines = 'noqa|^\s*(# )?<?https?://\S+>?$'
max-line-length = 90
max-nested-blocks = 3
msg-template="{path}:{module}:{line}: [{msg_id}({symbol}), {obj}] {msg}"
min-similarity-lines = 5
no-docstring-rgx = "^test_"
persistent = "no"

[tool.black]
line_length = 90

[tool.isort]
line_length = 90
multi_line_output = 3
include_trailing_comma = true

[build-system]
requires = ["poetry-core>=1.0.0a5"]
build-backend = "poetry.core.masonry.api"
```

**Comment by abn** [comment]
The short answer here is that this is expected and highlights the fragility of how editable installs are handled today. This is an expected side-effect of the transition to a more standard approach to handle editable installs in [PEP 517](https://www.python.org/dev/peps/pep-0517/) build backends as recommended in [PEP 610](https://www.python.org/dev/peps/pep-0610/). This does not have any functional change in behaviour, and it is important to note that curently there is no formal implemented standard around this.

Prior to #2400 (`1.1.0.a2`), `poetry install` would lead to the creation of `*.egg-link` files of the project being developed. The use of `*.egg-link` files has always been more of a hacky solution [[1](https://github.com/pypa/packaging-problems/issues/256#issuecomment-572278629), [2](https://www.python.org/dev/peps/pep-0610/#note-about-editable-installs)] that setuptools and by extension pip has relied on, , together with the absense of `*.dist-info` directory, to "detect" editable packages. This problem of communicating metadata relating to an editable install will go away once  the [PEP 610 specification](https://www.python.org/dev/peps/pep-0610/#specification) is fully implemented, in particular the following.

> When url refers to a local directory, the dir_info key MUST be present as a dictionary with the following key:
>
> editable (type: boolean): true if the distribution was installed in editable mode, false otherwise. If absent, default to false.
When url refers to a local directory, it MUST have the file sheme and be compliant with RFC 8089 [9]. In particular, the path component must be absolute. Symbolic links SHOULD be preserved when making relative paths absolute.

**Additional References:**
- https://discuss.python.org/t/pip-freeze-vcs-urls-and-pep-517-feat-editable-installs/1473/3

**Comment by abn** [comment]
> Would be good to have sorted since it's something that's blocking me from updating. I'd be happy to help with this if you need extra hands.

@snejus would be great to get some more details on how this is blocking your upgrade.

**Comment by snejus** [comment]
Okay, this makes sense. Though considering that many people do rely on editable installs, I wouldn't say that releasing these changes is the best option (even if we're talking about pre-releases). I take my words back if (a) majority of things going on now *depend on it being present* or (b) it's about to be completed. For example, maybe it would be possible to isolate PEP 610 to a specific _branch/letter of pre-releases_ (is that even a thing, really?), say `x.x.xbx`. At the same time, `x.x.xax` could include every update except PEP 610 work. Then people who don't care about editable installs can go ahead with `b`, while people that rely on them can stay with `a` and benefit from the rest of updates. 

I'm assuming here though that there is less coupling between things than there may actually be. Also, I've never seen a setup like this - probably because now I realise that it does sound like it'd be a pain in the ass to manage this kind of thing.

> @snejus would be great to get some more details on how this is blocking your upgrade.

For example, I often have `django`s `python manage.py runserver` running locally to make use of its auto-reload capabilities. 
Then, for some packages I run `mypy` on the package instead of paths. `pylint` also.

**Comment by abn** [comment]
To be clear, in this context, functionality has not changed.

The issue reported here has to do with how `pip` detects the "editable" package's location. When you run `poetry install` the current project's included directories are written to `<purelib>/python-<version>/site-packages/<package>.pth` file as described [here](https://docs.python.org/3/library/site.html). This means that your package source is loaded from the your source tree. This means commands like `poetry run python manage.py runserver` will run your current version of the code.

> In general, I find having the most recent code available in all contexts (i.e. editable mode) more preferable than not.

In development environments this may be true for most cases excluding isolated testing. This is not desireable nor recommended in managed environments.

**Comment by adamf** [comment]
This is also breaking our upgrade to 1.1.0 - we install a package in editable mode so we can, later in a docker build step, inject built protocol buffers into that package. Without editable mode this process breaks. Any ideas how to work around it in 1.1.0 series?

**Comment by finswimmer** [comment]
Hello @adamf,

the editable mode still works. The default behavior has changed. So if you define a path dependency, you have to set the `develop` key to `true` explicit:

```
my-package = {path = "../path/to/dependency", develop = true}
```

**Comment by nishantvarma** [comment]
@dpompeu-xgeeks, doesn't it have a problem? *remove* a package will "forget" its dependencies; so when you add again, you get the latest?

**Comment by dpompeu-xgeeks** [comment]
> @dpompeu-xgeeks, Removing a package will "forget" its dependencies, isn't that a problem? When you add again, you get the latest. Switching with minimal impacts will the best behaviour.

That's a good point. 

We got some changes in our `poetry.lock` file when we didn't want to: not sure if this was the reason , but it's possible. In the end, we pinned all library versions in `pyproject.toml` to guarantee there were no unexpected updates.

This works for us, but doesn't seem feasible for other use cases, eg a public library which should be more flexible with its dependencies versions.

**Comment by dpompeu-xgeeks** [comment]
Another workaround we used which worked more reliably, but with great effort:

1. develop changes in library
2. change the library's version to a temporary version, eg `0.5.1-experimental-d7ba`
2. `git tag 0.5.1-experimental-d7ba`
3. `git push --tag 0.5.1-experimental-d7ba`
4. in the main application, use `poetry add git+ssh://git@github.com/user/lib#0.5.1-experimental-d7ba`
5. test the changes 
6. repeat all previous steps if there was a bug

(We use git to host our internal libraries).

Being able to change existing dependency to editable mode would be much quicker.

**Comment by snejus** [comment]
@dpompeu-xgeeks have you tried specifying those dependencies as local paths? 

I have been successfully using the following configuration:
```toml
[tool.poetry.dependencies]
python = ">=3.6.2,<4"

beets = { path = "../beets", develop = true, extras = ["chroma", "discogs", "fetchart", "lastgenre", "mpdstats", "web"] }
beetcamp = { path = "../beetcamp", develop = true }
rich_tables = { path = "../rich-tables", develop = true }
beets-importreplace = { git = "https://github.com/edgars-supe/beets-importreplace.git" }
discogs-client = "*"
```

As expected, poetry uses a `.pth` file for each of the editable dependencies:
```sh
$ print -l $VIRTUAL_ENV/lib/python3.10/site-packages/*.pth(:t)
beetcamp.pth
beets.pth
distutils-precedence.pth
rich_tables.pth
_virtualenv.pth
```

**Comment by snejus** [comment]
@nishantvarma, in the case above (specifying a git-based dependency) inclusion of `develop = true` indeed makes the source code editable, however, `poetry` downloads the repository during the install and keeps it in your virtual environment (`$VIRTUAL_ENV/src/pendulum`) which is not ideal.

Instead of that I provide **a path** to a _local_ repository (e.g. `beetcamp = { path = "../beetcamp", develop = true }`). I do not need to tag or push the code to a remote repository in order to use it as a dependency in some other project which I am working on locally.

**Comment by dpompeu-xgeeks** [comment]
> @dpompeu-xgeeks have you tried specifying those dependencies as local paths?

The problem is when I change from git to local paths: poetry does not apply the changes.

I can reproduce this by creating a lib and app:

```bash
# mylib
$ echo 'print("Hello World from git tag")' > mylib.py
$ git commit -m "some message"
$ git rev-parse HEAD
cad773cb633c3d8be5ddd16ef6767d5d0bc6e7cc
$ git push
```

```bash
# myapp
$ poetry add git+ssh://git@github.com:dpompeu-xgeeks/example-lib.git#cad773cb633c3d8be5ddd16ef6767d5d0bc6e7cc
$ cat myapp.py
import mylib.mylib
$ poetry run python myapp.py 
Hello World from git tag
```

Now let's say I want to make some changes and debug them locally before pushing:

```bash
#mylib
$ echo 'print("Hello World from path")' > mylib.py
$ vim pyproject.toml # bump version, not sure if needed
```

```bash
#myapp
$ vim pyproject.toml # mylib = {path = "../mylib/", develop = true }
$ poetry lock
$ poetry install
$ poetry run python myapp.py
Hello World from git tag #  <--- no changes!!!
```

My expectation is that poetry would use the local files. 

Instead, I'm either removing the package before, or giving up and just pushing experimental versions to git, which is a slow process.

**Comment by dpompeu-xgeeks** [comment]
> @dpompeu-xgeeks,
> 
> Another possibility:
> 
> Start with `pendulum = "^2.1.2"`, `lock` and `install`. Update to the git with `develop = true`. Do a `poetry lock --no-update` to update the changes. Do an `rm -rf .venv` and `poetry install`. # Direct `poetry install` won't work for the same reason as above. This kind of a removal but preserves the lock.


@nishantvarma: Thanks for the tip, I can confirm this works! It does seem better than removing the packages.

However, I'd still hope this would work more simply: this way is prone to mistakes, eg forgetting to remove the package and then wondering why there are no changes. It has happened with different people in my project.

**Comment by snejus** [comment]
@dpompeu-xgeeks Could you confirm I understood it correctly:

1. Initially added the dependency through the git reference:
```$ poetry add git+ssh://git@github.com:dpompeu-xgeeks/example-lib.git#cad773cb633c3d8be5ddd16ef6767d5d0bc6e7cc```

2. Later removed the reference to the repository and replaced it with 
```mylib = {path = "../mylib/", develop = true }```

Could you also check what output do you see after running the following?

```
poetry --version
poetry shell
grep . $VIRTUAL_ENV/lib/*/*/*.pth | grep -oP '\K/lib.*'
```

**Comment by snejus** [comment]
I've tried doing the same in my machine and encountered your issue. After some investigation, it seems that `poetry` does not manage to uninstall the outdated (git) version of your dependency which then conflicts with your editable install later. 

To fix this, **before** adding the path dependency, run both `poetry remove mylib` and `pip uninstall mylib` to cleanly get rid of it from your environment. You can double-check with `poetry run python myapp.py` - it should fail with `ModuleNotFoundError: No module named 'mylib'`.

Then, add the local path dependency and lock/install. It worked okay for me.

**Comment by dpompeu-xgeeks** [comment]
FYI this seems to be fixed by the recent release `1.2.0`.
